### PR TITLE
Support repline 0.4

### LINKED ISSUE #651: hnix-0.9.0 failing to build (broken by repline-0.3.0.0?)
We are seeing the following build failure on stackage nightly which may have been caused by the recent release of `repline-0.3.0.0`. We have suspended hnix from nightly but let us know when the issue has been resolved and we will restore it.

---
```
    Configuring hnix-0.9.0...
    Preprocessing library for hnix-0.9.0..
    Building library for hnix-0.9.0..
    Preprocessing executable 'hnix' for hnix-0.9.0..
    Building executable 'hnix' for hnix-0.9.0..
    [2 of 3] Compiling Repl [System.Console.Repline changed]

    /var/stackage/work/unpack-dir/unpacked/hnix-0.9.0-3049e4182aff661b1e01927a219766f0adee953b95bfa4e67fd33d217a9ea7fb/main/Repl.hs:61:7: error:
        • Couldn't match expected type ‘StateT (IState t0 f0 m0) m ()’
                      with actual type ‘HaskelineT (StateT (IState t2 f2 m3) m3) a0
                                        -> HaskelineT (StateT (IState t2 f2 m3) m3) ExitDecision
                                        -> StateT (IState t2 f2 m3) m3 ()’
        • Probable cause: ‘evalRepl’ is applied to too few arguments
          In the second argument of ‘($)’, namely
            ‘evalRepl
               (return prefix) cmd options (Just ':') completer welcomeText’
          In the expression:
            flip evalStateT initState
              $ evalRepl
                  (return prefix) cmd options (Just ':') completer welcomeText
          In an equation for ‘main’:
              main
                = flip evalStateT initState
                    $ evalRepl
                        (return prefix) cmd options (Just ':') completer welcomeText
                where
                    prefix = "hnix> "
                    welcomeText
                      = liftIO
                          $ putStrLn
                              $ "Welcome to hnix "
                                  <> showVersion version <> ". For help type :help\n"
        • Relevant bindings include
            prefix :: HaskelineT (StateT (IState t2 f2 m3) m3) String
              (bound at main/Repl.hs:63:3)
            main :: m () (bound at main/Repl.hs:60:1)
       |
    61 |     $ evalRepl (return prefix) cmd options (Just ':') completer welcomeText
       |       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    /var/stackage/work/unpack-dir/unpacked/hnix-0.9.0-3049e4182aff661b1e01927a219766f0adee953b95bfa4e67fd33d217a9ea7fb/main/Repl.hs:61:36: error:
        • Couldn't match type ‘[Char]’ with ‘Char’
          Expected type: System.Console.Repline.Options
                           (HaskelineT (StateT (IState t2 f2 m3) m3))
            Actual type: [(String, [String] -> Repl e0 t2 f2 m3 ())]
        • In the third argument of ‘evalRepl’, namely ‘options’
          In the second argument of ‘($)’, namely
            ‘evalRepl
               (return prefix) cmd options (Just ':') completer welcomeText’
          In the expression:
            flip evalStateT initState
              $ evalRepl
                  (return prefix) cmd options (Just ':') completer welcomeText
       |
    61 |     $ evalRepl (return prefix) cmd options (Just ':') completer welcomeText
       |                                    ^^^^^^^

    /var/stackage/work/unpack-dir/unpacked/hnix-0.9.0-3049e4182aff661b1e01927a219766f0adee953b95bfa4e67fd33d217a9ea7fb/main/Repl.hs:61:55: error:
        • Couldn't match expected type ‘Maybe String’
                      with actual type ‘CompleterStyle (StateT (IState t1 f1 m1) m1)’
        • In the fifth argument of ‘evalRepl’, namely ‘completer’
          In the second argument of ‘($)’, namely
            ‘evalRepl
               (return prefix) cmd options (Just ':') completer welcomeText’
          In the expression:
            flip evalStateT initState
              $ evalRepl
                  (return prefix) cmd options (Just ':') completer welcomeText
       |
    61 |     $ evalRepl (return prefix) cmd options (Just ':') completer welcomeText
       |                                                       ^^^^^^^^^

    /var/stackage/work/unpack-dir/unpacked/hnix-0.9.0-3049e4182aff661b1e01927a219766f0adee953b95bfa4e67fd33d217a9ea7fb/main/Repl.hs:61:65: error:
        • Couldn't match kind ‘* -> *’ with ‘*’
          When matching types
            m2 :: * -> *
            CompleterStyle :: (* -> *) -> *
          Expected type: CompleterStyle (StateT (IState t2 f2 m3) m3)
            Actual type: m2 ()
        • In the sixth argument of ‘evalRepl’, namely ‘welcomeText’
          In the second argument of ‘($)’, namely
            ‘evalRepl
               (return prefix) cmd options (Just ':') completer welcomeText’
          In the expression:
            flip evalStateT initState
              $ evalRepl
                  (return prefix) cmd options (Just ':') completer welcomeText
        • Relevant bindings include
            welcomeText :: m2 () (bound at main/Repl.hs:64:3)
       |
    61 |     $ evalRepl (return prefix) cmd options (Just ':') completer welcomeText
       |                                                                 ^^^^^^^^^^^
```

### PR DESCRIPTION
**Pull request by sorki**
Closes #651.

`haskeline` override needs `dontCheck` as it tries to run an executable.

Tested build and repl itself.

### CODE REVIEW

File: main/Repl.hs
Comment (sorki) [inline main/Repl.hs]: This can be used to enable multi-line support which could be neat for our repl.

**Review by sjakobi — APPROVED** [review]
Nice! :)

**Comment by sorki** [comment]
Almost! I might have picked too recent `haskell-updates` branch which is probably not part of unstable yet.

**Comment by Anton-Latukha** [comment]
Just basically convert to PRs to draft to indicate the state.

CI works as normal in draft form.

**Comment by Anton-Latukha** [comment]
@sorki, @sjakobi overall I am happy we have so good team, real improvements started to show up. We together can figure it out over the time.

**Comment by sorki** [comment]
```
Welcome to hnix 0.9.0. For help type :help

hnix> :paste
-- Entering multi-line mode. Press <Ctrl-D> to finish.
| { just
| =
| ''
| testing
| '';
| }
|
{ just = "<CYCLE>"; }
hnix>
Goodbye.
```

..also refactored imports a bit. `Repline 0.4` should land in `nixpkgs-unstable` soon I guess, keeping as draft as I might add better help (similar to Dhalls `HelpOption` one which is neat) and few exception handlers. Dhalls Unicode also looks interesting, what do you think about `λ> ` prompt? :smile_cat:

**Comment by sjakobi** [comment]
@sorki _Very_ nice! :)


> what do you think about `λ> ` prompt?

Personally, I like the current `hnix>` "branding", but it's not very important to me.

**Comment by sorki** [comment]
Another sample, notice how evaluating `c` makes next `:browse` show values as well.

```
hnix> a = 2
{ a = "<CYCLE>"; }
hnix> b = 3
{ b = "<CYCLE>"; }
hnix> c = a + b
{ c = "<CYCLE>"; }
hnix> :browse
a = "<CYCLE>"
b = "<CYCLE>"
c = "<CYCLE>"
hnix> :t c
an integer
hnix> c
5
hnix> :browse
a = 2
b = 3
c = 5
hnix> :help
Available commands:

:help 
    Print help text
:load FILENAME
    Load .nix file into scope
:browse 
    Browse bindings in interpreter context
:type EXPRESSION
    Evaluate expression or binding from context and print the type of the result value
:quit 
    Quit interpreter
:debug 
    Enable REPL debugging output
```

I would like to know your opinion about `parseExprOrBinding`.

It also occurred to me that instead of `:debug` it would be better to add `:set` command just like GHCI one, so we can do e.g. `:set strict`, `:set debug`, `:set prompt "🐈> "` which leads me to something like `.hnix-repl-rc`. I'll poke this later as it seems easy since we already track state with `IState`.

**Comment by sorki** [comment]
`callHackage` won't help either as all-cabal-hashes is pinned as well heh. I'm using `nixpkgs @  5db4111908ffd59313b9e7879e177088743d84ca` and `nix-build --argstr compiler ghc8101` if you want to test.

**Comment by sjakobi** [comment]
> I would like to know your opinion about `parseExprOrBinding`.

https://github.com/haskell-nix/hnix/blob/1c77ef8bdd2dc6742afd4f7793f31b9a4748cda9/main/Repl.hs#L164-L175

I don't quite understand the motivation. The comment indicates that its just a shorthand for writing a set. Is that very useful?

_Note:_ I'm a Nix noob. I have hardly any experience with Nix REPL workflows.

If the goal is to make this binding available in the context, that seems more useful to me. I think I wouldn't do it this way in `dhall repl` though, because it could confuse users about Dhall syntax. I'm happy with the `:let` command that we have there. `hnix` is a different project though and possibly targeted at more experienced users who already know the Nix syntax well.

**Comment by sorki** [comment]
> If the goal is to make this binding available in the context, that seems more useful to me. I think I wouldn't do it this way in `dhall repl` though, because it could confuse users about Dhall syntax. I'm happy with the `:let` command that we have there. `hnix` is a different project though and possibly targeted at more experienced users who already know the Nix syntax well.

Exactly. It's just a shorthand that takes a binding e.g. `a = 2` and creates a singleton attrset out of it as `{ a = 2; }` so it can be (edit: actually parsed and) used as part of the REPL context. I've noticed the `:let` command in Dhall and it could be replicated here as well but I like this shorthand functionality which saves typing all the `:let`s and mimics current `nix repl`.

**Comment by sjakobi** [comment]
> mimics current `nix repl`.

Well that's a good point! :+1:

**Comment by sorki** [comment]
Diverged a bit from `repline 0.4` compatibility but it's now in much better shape overall.

- Removed overrides
- Reworked tab completion

File: main/Main.hs
Comment (sjakobi) [inline main/Main.hs]: I think a `do` expression might be slightly more readable here.

(I general I feel that some of the code in this project is so dense that it hurts readability. I'm still very new to it though – maybe I just need to get used to it.)

**Comment by sjakobi** [comment]
I just noticed that you marked this as ready for review. GitHub somehow hadn't sent me a notification for that. I'll try to review this tomorrow (Friday).

File: main/Repl.hs
Comment (sjakobi) [inline main/Repl.hs:124]: Does this mean that commands can be abbreviated? E.g. `:l` instead of `:load`?

File: main/Repl.hs
Comment (sjakobi) [inline main/Repl.hs:113]: Could you include a description of the adjustment here?

File: main/Repl.hs
Comment (sjakobi) [inline main/Repl.hs]: Aside: Do you happen to know why we disable these warnings?

**Review by sjakobi — APPROVED** [review]
Wow. That looks like a lot of new features! :)

I definitely don't understand everything that's going on here. I'll report any issues once I start using it though. ;)

**Comment by sjakobi** [comment]
Oh, a bit of documentation of the REPL features would be great to have! Maybe we could just show a few examples in the README?!

File: main/Repl.hs
Comment (sorki) [reply main/Repl.hs:124]: Yup, that's what repline allows by itself as well.

File: main/Main.hs
Comment (sorki) [reply main/Main.hs]: Done. I agree with the density/readability, so far I've only contributed one such fix -  we can slowly improve this as we progress through the code base.

File: main/Repl.hs
Comment (sorki) [reply main/Repl.hs:113]: Done, added a TODO for myself to file a PR exposing this from `repline` or adding support for `.rc` directly.

File: main/Repl.hs
Comment (sorki) [reply main/Repl.hs]: Nice catch, missed that completely - removed and cleaned up leftover imports.

**Comment by sorki** [comment]
> Oh, a bit of documentation of the REPL features would be great to have! Maybe we could just show a few examples in the README?!

I was wondering what's the good place for that - added a short README section via 9d7d0fd8de2d833c1952e97037b149f291fbf80b mostly to document propagating `--eval -E` as `input`.

I'll create a test and Asciinema demo later using my new toy - https://github.com/sorki/polytype/

File: README.md
Comment (sjakobi) [inline README.md]:
```suggestion
## Using the REPL
```

File: README.md
Comment (sjakobi) [inline README.md]:
```suggestion
To enter the `hnix` REPL use
```

File: README.md
Comment (sjakobi) [inline README.md]:
```suggestion
To evaluate an expression and make it available in the REPL
as the `input` variable use
```

File: README.md
Comment (sjakobi) [inline README.md]:
```suggestion
Use the `:help` command for a list of all available REPL commands.
```

File: main/Repl.hs
Comment (sjakobi) [reply main/Repl.hs:113]: Thanks! :)

File: main/Repl.hs
Comment (sjakobi) [inline main/Repl.hs]: The `Maybe` doesn't look strictly necessary, or is it?

```suggestion
    evalText :: (MonadNix e t f m) => Text -> m (NValue t f m)
```

File: main/Repl.hs
Comment (sjakobi) [inline main/Repl.hs]: I think this could be simpler, especially when `evalText` doesn't return a `Maybe`. Maybe something like this?

```suggestion
    (Data.HashMap.Lazy.fromList $ ("builtins", builtins) : fmap ("input",) (Data.Maybe.maybeToList mIni))
```

File: main/Repl.hs
Comment (sjakobi) [inline main/Repl.hs]:
```suggestion
  -- Attributes of sets in REPL context
```

**Review by sjakobi — APPROVED** [review]
I have a few more minor comments. If they are too detailed or not helpful in other ways, just point it out, and I'll try to adjust my future reviews.

> > Oh, a bit of documentation of the REPL features would be great to have! Maybe we could just show a few examples in the README?!
> 
> I was wondering what's the good place for that - added a short README section via [9d7d0fd](https://github.com/haskell-nix/hnix/commit/9d7d0fd8de2d833c1952e97037b149f291fbf80b) mostly to document propagating `--eval -E` as `input`.

I think the README is a good place. It's probably the first place where people would look for documentation. If the readme gets very long, we could add a table of contents.

> I'll create a test and Asciinema demo later using my new toy - https://github.com/sorki/polytype/

Intriguing! :)

File: main/Repl.hs
Comment (sorki) [reply main/Repl.hs]: Yup, this was only to go nicely with `catMaybes`.

File: main/Repl.hs
Comment (sorki) [reply main/Repl.hs]: Thanks, TIL `TupleSections` & `maybeToList` :+1:

**Comment by sorki** [comment]
> I have a few more minor comments. If they are too detailed or not helpful in other ways, just point it out, and I'll try to adjust my future reviews.

Thank you for the review! I don't mind fixing nitpicks or style issues, in case of README I totally forgot to enable spellchecking.

File: main/Repl.hs
Comment (sjakobi) [inline main/Repl.hs]: The last puzzler I'm going to point out. ;)

I think it could be written roughly like this:
```suggestion
  for_ mVal $ \val -> do
    s <- lift . lift . showValueType $ val
    liftIO $ putStrLn s
```

**Review by sjakobi — APPROVED** [review]
One very very last comment! ;)

File: main/Repl.hs
Comment (sorki) [reply main/Repl.hs]: Neat! Thank you!

**Comment by Anton-Latukha** [comment]
So, looks like the consensus is reached.

I had a brief look, and it looks fine. Of course, it also expands usage and works, so...

Do you consider it ready to merge?

**Comment by sorki** [comment]
> So, looks like the consensus is reached.
> 
> I had a brief look, and it looks fine. Of course, it also expands usage and works, so...
> 

Thanks!

> Do you consider it ready to merge?

Yes, I think it's good to go.

**Comment by sjakobi** [comment]
FYI, I usually don't merge other committers' PRs because people have different preferences regarding the commit style.

When I approve and leave a comment, I mean to express that the PR is good to merge after addressing the comment.

**Comment by Anton-Latukha** [comment]
Well )

**Comment by Anton-Latukha** [comment]
I started to use "after merge" and "after report closed" for tangential thread discussions. Some discussions regarding the project are great to move to the global project chat, but the local unrelated discussions are great to continue in closed threads.

I personally use Merge as is and commits. But I always wait to be persuaded and to learn more. You know I love that "Create merge commit" button.

And I have/had my arguments about it.

And nobody so far stated arguments for the weight to pull me to the other side.

But yesterday, at last, I stumbled upon a lot of good solid arguments towards squash&merge:
https://github.com/marketplace/actions/autosquash#why-squash-and-merge.

**Comment by Anton-Latukha** [comment]
To be honest, I always knew the both approaches have strong argumentation and there is a reason people stick to one or another.

My response - is to why not have the abilities of both.

For me - the strongest argument about/between merge systems is the `git bisect`, and it is a syndrome for me. Both systems argue, but in real - they both failed to arrive at the solution.

With "Create merge commit" it is too much granularity, `git bisect` probably, and most probably would fail in some non-atomic commits.
With Squash&merge - is not enough granularity.

For example, if we Squash&merge this PR - how someone would find the particular point that caused the issue in an epic commit. Parts of the system was not even present before it, any person can try to debug the behavior and they only have this whole single commit that submitted subsystem as a whole.

"Merge commit" has metadata information that allows it to treat it as a Squash&merge commit.

So, if we have a mode that allows to treat all linked commits to the Merge commit as a single abstraction - we get the atomicity of the Squash&merge, and the ability of further granularity inside a PR and the whole history.

So, in simplified form: if we are able to `git bisect` over the "Merge commits" - we are able to have a cake and eat it too.

**Comment by Anton-Latukha** [comment]
So, the proper response is: Created the: https://github.com/Anton-Latukha/merge-bisect project to address this.

The simple shell script so far seems the most balanced approach that is also approachable for people to understand contribute, for this case at this stage.

And there is also a Python version https://github.com/dealertrack/merge-bisect/.

**Comment by sjakobi** [comment]
What I like about squashed PRs:

* There's less noise in the commit history.
* The commit history is closer to my own memory of the development history: I mostly remember PRs, but I'm less aware of the individual commits that go into a PR. That makes it easier for me to navigate the commit history.
* Each commit includes a reference to the PR, so I can more easily find the discussion that went into a commit.
* The chronological commit history is less confusing and shows commits in the order in which they got into the `master` branch.

I haven't really had any problems with giant commits, possibly because I'm mostly involved in projects where giant PRs don't happen very often.

**Comment by Anton-Latukha** [comment]
Well, you are right that the small PRs are the best approach.

To the Squash & merge the diametral approach is Rebase and merge. And Merge commits, in real, are the middle way that allows to have the flexibility of both.

I am always for extending the tooling. If you do not want to see something - alias the `git log` to not show not `master` branch commits and here it is. You would never see the internal PR commits not only in this repository, but in all repositories and workflows you would ever be part of.

I am on the contrary. I never look at the Git history until I need to, and when I need to - it is an automatic `git bisect`, OR `git blame` where I prefer small atomic steps and according to them comments, which are preserved in Merge commit or Rebase merge.

Sorki wrote a great commit comment to every commit, why squash all that.

+====

In repo I formed the full algorithm.

Maybe I need to bite the bullet and write it in Haskell.

**Comment by sorki** [comment]
I was considering squashing this into one commit but decided against it since the commits create a logical progression of work (migration fix + separate features) during few days of working on it while waiting for `unstable`. They should even build separately but I might have screwed that with further fix-ups. Squashing is easy but exactly like you point out - the history is lost forever :)

`merge-bisect` looks cool! 

Btw if we do a lot of refactoring / style / imports changes it might be worth to implement `.git-blame-ignore-revs` file listing these noisy commits since git supports it via

```
git blame --ignore-revs-file .git-blame-ignore-revs file
```

or via `blame.ignoreRevFile` config option.

**Pull request merged**
